<h2>Decision Trees</h2>

We used this classification algorithm to build a model from historical data of patients, and their response to different medications. Then we used  decision tree to predict the class of an unknown patient, or to find a proper drug for a new patient.

In [1]:
import numpy as np 
import pandas as pd

<h4>About the dataset</h4>

The feature sets of this dataset are Age, Sex, Blood Pressure, and Cholesterol of patients, and the target is the drug that each patient responded to one of 5 medications, Drug A, Drug B, Drug c, Drug x and y.It is a sample of binary classifier, and we can use the training part of the dataset to build a decision tree, and then use it to predict the class of a unknown patient, or to prescribe it to a new patient.

In [17]:
filepath='https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/drug200.csv'

In [18]:
df = pd.read_csv(filepath)
df.head()

Age Sex      BP Cholesterol  Na_to_K   Drug
0   23   F    HIGH        HIGH   25.355  drugY
1   47   M     LOW        HIGH   13.093  drugC
2   47   M     LOW        HIGH   10.114  drugC
3   28   F  NORMAL        HIGH    7.798  drugX
4   61   F     LOW        HIGH   18.043  drugY

 <h3>Pre-processing</h3>

In [23]:
x = df[['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K']].values
x[0:3]

array([[23, 'F', 'HIGH', 'HIGH', 25.355],
       [47, 'M', 'LOW', 'HIGH', 13.093],
       [47, 'M', 'LOW', 'HIGH', 10.113999999999999]], dtype=object)

some features in this dataset are categorical such as Sex or BP.  Sklearn Decision Trees do not handle categorical variables. But we can convert these features to numerical values.

In [20]:
from sklearn.preprocessing import LabelEncoder

In [24]:
le_sex=LabelEncoder()
le_sex.fit(['F','M'])
x[:,1]=le_sex.transform(x[:,1])

le_bp=LabelEncoder()
le_bp.fit([ 'LOW', 'NORMAL', 'HIGH'])
x[:,2]=le_bp.transform(x[:,2])


le_chol=LabelEncoder()
le_chol.fit(['NORMAL','HIGH'])
x[:,3]=le_chol.transform(x[:,3])

x[0:3]

array([[23, 0, 0, 0, 25.355],
       [47, 1, 1, 0, 13.093],
       [47, 1, 1, 0, 10.113999999999999]], dtype=object)

In [26]:
y = df["Drug"]

<h3>Train/Test split</h3>

In [28]:
from sklearn.model_selection import train_test_split

In [29]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=3)

<h3>Modeling</h3>

I created an instance of the DecisionTreeClassifier called drugTree.
Inside of the classifier, specify criterion="entropy" so I can see the information gain of each node.

In [32]:
from sklearn.tree import DecisionTreeClassifier

In [33]:
drugTree = DecisionTreeClassifier(criterion="entropy", max_depth = 4)

In [35]:
drugTree.fit(x_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

<h3>Prediction</h3>

In [41]:
yhat = drugTree.predict(x_test)


compare y and yhat(predicted values with real values in testset)

In [42]:
print (yhat[0:5])
print (y_test[0:5])

['drugY' 'drugX' 'drugX' 'drugX' 'drugX']
40     drugY
51     drugX
139    drugX
197    drugX
170    drugX
Name: Drug, dtype: object


<h3>Evaluation</h3>

In [40]:
from sklearn.metrics import accuracy_score

In [43]:
print("DecisionTrees's Accuracy: ",accuracy_score(y_test, yhat))

DecisionTrees's Accuracy:  0.9833333333333333


<h3>Conclusion</h3>

with this model we can predict 0.98  class of a new patiant correctly